File: apps/desktop/src/lib/tauri/proxy.ts
import type {
	CreateProxyInput,
	ProfileProxyAssignment,
	ProxyAssignment,
	ProxyCheckResult,
	Proxy as ProxyRecord,
	TestProxyInput,
	UpdateProxyInput,
} from "@/types/proxy";

import { invokeCommand } from "./client";

export const proxyApi = {
	list() {
		return invokeCommand<ProxyRecord[]>("proxy_list");
	},

	get(id: string) {
		return invokeCommand<ProxyRecord>("proxy_get", { id });
	},

	create(input: CreateProxyInput) {
		return invokeCommand<ProxyRecord>("proxy_create", { input });
	},

	update(id: string, input: UpdateProxyInput) {
		return invokeCommand<ProxyRecord>("proxy_update", { id, input });
	},

	archive(id: string) {
		return invokeCommand<void>("proxy_archive", { id });
	},

	check(id: string) {
		return invokeCommand<ProxyCheckResult>("proxy_check", { id });
	},

	testInput(input: TestProxyInput) {
		return invokeCommand<ProxyCheckResult>("proxy_test_input", { input });
	},

	assign(profileId: string, proxyId: string) {
		return invokeCommand<void>("proxy_assign", { profileId, proxyId });
	},

	unassign(profileId: string) {
		return invokeCommand<void>("proxy_unassign", { profileId });
	},

	getProfileAssignment(profileId: string) {
		return invokeCommand<ProfileProxyAssignment>(
			"proxy_get_profile_assignment",
			{
				profileId,
			},
		);
	},

	listAssignments(proxyId: string) {
		return invokeCommand<ProxyAssignment[]>("proxy_list_assignments", {
			proxyId,
		});
	},

	listChecks(proxyId: string, limit?: number) {
		return invokeCommand<ProxyCheckResult[]>("proxy_list_checks", {
			proxyId,
			limit,
		});
	},
};

File: apps/desktop/src-tauri/src/commands/proxy.rs
use tauri::State;

use crate::domain::proxy::{
    CreateProxyInput, ProfileProxyAssignmentDto, ProxyAssignmentDto, ProxyCheckResultDto, ProxyDto,
    TestProxyInput, UpdateProxyInput,
};
use crate::error::AppError;
use crate::state::AppState;

#[tauri::command]
pub async fn proxy_list(state: State<'_, AppState>) -> Result<Vec<ProxyDto>, AppError> {
    state.proxy_service.list(&state).await
}

#[tauri::command]
pub async fn proxy_get(state: State<'_, AppState>, id: String) -> Result<ProxyDto, AppError> {
    state.proxy_service.get(&state, &id).await
}

#[tauri::command]
pub async fn proxy_create(
    state: State<'_, AppState>,
    input: CreateProxyInput,
) -> Result<ProxyDto, AppError> {
    state.proxy_service.create(&state, input).await
}

#[tauri::command]
pub async fn proxy_update(
    state: State<'_, AppState>,
    id: String,
    input: UpdateProxyInput,
) -> Result<ProxyDto, AppError> {
    state.proxy_service.update(&state, &id, input).await
}

#[tauri::command]
pub async fn proxy_archive(state: State<'_, AppState>, id: String) -> Result<(), AppError> {
    state.proxy_service.archive(&state, &id).await
}

#[tauri::command]
pub async fn proxy_check(
    state: State<'_, AppState>,
    id: String,
) -> Result<ProxyCheckResultDto, AppError> {
    state.proxy_service.check(&state, &id).await
}

#[tauri::command]
pub async fn proxy_test_input(
    state: State<'_, AppState>,
    input: TestProxyInput,
) -> Result<ProxyCheckResultDto, AppError> {
    state.proxy_service.test_input(&state, input).await
}

#[tauri::command]
pub async fn proxy_assign(
    state: State<'_, AppState>,
    profile_id: String,
    proxy_id: String,
) -> Result<(), AppError> {
    state.proxy_service.assign(&state, &profile_id, &proxy_id).await
}

#[tauri::command]
pub async fn proxy_unassign(
    state: State<'_, AppState>,
    profile_id: String,
) -> Result<(), AppError> {
    state.proxy_service.unassign(&state, &profile_id).await
}

#[tauri::command]
pub async fn proxy_get_profile_assignment(
    state: State<'_, AppState>,
    profile_id: String,
) -> Result<ProfileProxyAssignmentDto, AppError> {
    state.proxy_service.get_profile_assignment(&state, &profile_id).await
}

#[tauri::command]
pub async fn proxy_list_assignments(
    state: State<'_, AppState>,
    proxy_id: String,
) -> Result<Vec<ProxyAssignmentDto>, AppError> {
    state.proxy_service.list_assignments(&state, &proxy_id).await
}

#[tauri::command]
pub async fn proxy_list_checks(
    state: State<'_, AppState>,
    proxy_id: String,
    limit: Option<u32>,
) -> Result<Vec<ProxyCheckResultDto>, AppError> {
    state
        .proxy_service
        .list_checks(&state, &proxy_id, limit.unwrap_or(20))
        .await
}
